src/typechecker.ts

```diff
@@ -298,3 +357,397 @@ function textFromPreviousPart(seq: Seq, index: number): string | null {
+    // @ts-expect-error
+    /^\s+$/.test(items[items.length - 1].contents)
+  ) {
+    items.pop();
+  }
+  return items;
+}
+
+type Type =
+  | { kind: 'unknown' } // top
+  | { kind: 'never' } // bottom
+  | { kind: 'union'; of: NonUnion[] } // constraint: nothing in the union dominates anything else in the union
+  | { kind: 'list'; of: Type }
+  | { kind: 'record' } // TODO more precision
+  | { kind: 'completion'; of: Type }
+  | { kind: 'real' }
+  | { kind: 'integer' }
+  | { kind: 'non-negative integer' }
+  | { kind: 'concrete real'; value: string }
+  | { kind: 'ES value' }
+  | { kind: 'string' }
+  | { kind: 'number' }
+  | { kind: 'integral number' }
+  | { kind: 'bigint' }
+  | { kind: 'null' }
+  | { kind: 'concrete string'; value: string }
+  | { kind: 'concrete number'; value: string }
+  | { kind: 'concrete bigint'; value: string }
+  | { kind: 'enum value'; value: string };
+
+type NonUnion = Exclude<Type, { kind: 'union' }>;
+
+const simpleKinds = new Set<Type['kind']>([
+  'unknown',
+  'never',
+  'record',
+  'real',
+  'integer',
+  'non-negative integer',
+  'ES value',
+  'string',
+  'number',
+  'integral number',
+  'bigint',
+]);
+
+const dominateGraph: Partial<Record<Type['kind'], Type['kind'][]>> = {
+  // @ts-expect-error TS does not know about __proto__
+  __proto__: null,
+  record: ['completion'],
+  real: ['integer', 'non-negative integer', 'concrete real'],
+  integer: ['non-negative integer'],
+  'ES value': [
+    'string',
+    'number',
+    'integral number',
+    'bigint',
+    'concrete string',
+    'concrete number',
+    'concrete bigint',
+  ],
+  string: ['concrete string'],
+  number: ['integral number', 'concrete number'],
+  bigint: ['concrete bigint'],
+};
+
+/*
+The type lattice used here is very simple (aside from explicit unions).
+As such we really only need to define the `dominates` relationship and apply trivial rules:
+if `x` dominates `y`, then `join(x,y) = x` and `meet(x,y) = y`; if neither dominates the other than the join is top and the meet is bottom.
+Unions take a little more work.
+*/
+
+function dominates(a: Type, b: Type): boolean {
+  if (a.kind === 'unknown' || b.kind === 'never') {
+    return true;
+  }
+  if (b.kind === 'union') {
+    return b.of.every(t => dominates(a, t));
+  }
+  if (a.kind === 'union') {
+    // not necessarily true for arbitrary lattices, but true for ours
+    return a.of.some(t => dominates(t, b));
+  }
+  if (
+    (a.kind === 'list' && b.kind === 'list') ||
+    (a.kind === 'completion' && b.kind === 'completion')
+  ) {
+    return dominates(a.of, b.of);
+  }
+  if (simpleKinds.has(a.kind) && simpleKinds.has(b.kind) && a.kind === b.kind) {
+    return true;
+  }
+  if (dominateGraph[a.kind]?.includes(b.kind) ?? false) {
+    return true;
+  }
+  if (
+    (a.kind === 'integer' && b.kind === 'concrete real') ||
+    (a.kind === 'integral number' && b.kind === 'concrete number')
+  ) {
+    return !b.value.includes('.');
```

Comment:
`1.23456789e+23` is integral and a canonical ToString. 

```suggestion
    let n = parseFloat(b.value);
    return n === Math.round(n);
```

Alternatively, you could just keep a number instead of a string.

Reply:
Do we actually want to allow people to write numbers like `*1.23456789e+23*<sub>𝔽</sub>`?

Reply:
I guess I would slightly lean toward only permitting `DecimalDigits` and `DecimalDigits . DecimalDigits` and requiring that they be an exactly-representable float value. This is an editorial question we should confirm at an editor call.

Regardless though, I would still change the representation to `number` and move the complexity of what's permitted upstream.

Reply:
1 is exactly representable though...

Reply:
Pretend I said `0.1` like I obviously meant.

Reply:
That could be fine if we add some explanation in the Notational Conventions section saying that we mean the nearest-representable float. This is a topic for editor call.

Reply:
Also, changing this to be a `number` doesn't work for reals, which can in principle be values which are not representable as doubles. So I would still need to keep the string-handling logic for the real case. Given that, I don't want to introduce a different path just four doubles. That's a lot of extra complexity for no benefit.

Reply:
You already have different paths for reals and doubles.

Reply:
In some places, but not in this case.